src/diffusers/training_utils.py

```diff
@@ -39,88 +40,173 @@ def set_seed(seed: int):
     # ^^ safe to call this function even if cuda is not available
 
 
+# Adapted from torch-ema https://github.com/fadel/pytorch_ema/blob/master/torch_ema/ema.py#L14
 class EMAModel:
     """
     Exponential Moving Average of models weights
     """
 
     def __init__(
         self,
-        model,
-        update_after_step=0,
-        inv_gamma=1.0,
-        power=2 / 3,
-        min_value=0.0,
-        max_value=0.9999,
-        device=None,
+        parameters: Iterable[torch.nn.Parameter],
+        decay: float = 0.9999,
+        update_after_step: int = 0,
+        use_ema_warmup: bool = False,
+        inv_gamma: Union[float, int] = 1.0,
+        power: Union[float, int] = 2 / 3,
     ):
         """
+        Args:
+            parameters (Iterable[torch.nn.Parameter]): The parameters to track.
+            decay (float): The decay factor for the exponential moving average.
+            update_after_step (int): The number of steps to wait before starting to update the EMA weights.
+            use_ema_warmup (bool): Whether to use EMA warmup.
+            inv_gamma (float):
+                Inverse multiplicative factor of EMA warmup. Default: 1. Only used if `use_ema_warmup` is True.
+            power (float): Exponential factor of EMA warmup. Default: 2/3. Only used if `use_ema_warmup` is True.
+
         @crowsonkb's notes on EMA Warmup:
             If gamma=1 and power=1, implements a simple average. gamma=1, power=2/3 are good values for models you plan
             to train for a million or more steps (reaches decay factor 0.999 at 31.6K steps, 0.9999 at 1M steps),
             gamma=1, power=3/4 for models you plan to train for less (reaches decay factor 0.999 at 10K steps, 0.9999
             at 215.4k steps).
-
-        Args:
-            inv_gamma (float): Inverse multiplicative factor of EMA warmup. Default: 1.
-            power (float): Exponential factor of EMA warmup. Default: 2/3.
-            min_value (float): The minimum EMA decay rate. Default: 0.
         """
+        parameters = list(parameters)
+        self.shadow_params = [p.clone().detach() for p in parameters]
```

Comment:
The class expected a model, but now it expects a list of parameters. Can we somehow avoid this breaking change?

Reply:
Aah, good catch! IMO it's safer to just pass `parameters`,which makes it compatible with multi-gpu, deepspeed etc.
Think as @patrickvonplaten said, the breaking change should be fine.

Reply:
Think if we can avoid a breaking change, let's try to do it though no? Should be relatively easy by doing:
```suggestion
        if issubclass(parameters, torch.nn.Module):
            deprecate(...)
            parameters = parameters.parameters()
            
        parameters = list(parameters)
        self.shadow_params = [p.clone().detach() for p in parameters]
```

Reply:
That should catch 99% of the breaking use cases no? We could also add a kwargs.pop("model") to be super exact